src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,68 +342,116 @@ private static void assertVersionMatch() {
             int attempt = 0;
             // We keep doing this until we can create something new or fail badly
             while (attempt < RETRY_LIMIT) {
                 attempt++;
-                if (rndStream.read(rndBytes) != rndBytes.length) {
-                    throw new AssertionError("Unable to read enough entropy");
+                if (rndStream.read(rndBytes) == rndBytes.length) {
+                    return rndBytes;
                 }
+            }
+        }
+        throw new AssertionError("Unable to read enough entropy");
+    }
 
-                final StringBuilder fileName = new StringBuilder(prefix);
-                for (byte b : rndBytes) {
-                    // We convert to an unsigned integer first to avoid sign-bit extension when converting to hex.
-                    String hexByte = Integer.toHexString(Byte.toUnsignedInt(b));
-                    if (hexByte.length() == 1) {
-                        fileName.append('0');
-                    }
-                    fileName.append(hexByte);
-                }
-                fileName.append(suffix);
-
-                final Path tmpFile = tmpDir.resolve(fileName.toString());
-
-                try {
-                    final Path result = Files.createFile(tmpFile, permissions);
-                    if (DebugFlag.VERBOSELOGS.isEnabled()) {
-                        LOG.log(Level.FINE, "Created temporary library file after " + attempt + " attempts");
-                    }
-                    result.toFile().deleteOnExit();
-                    return result;
-                } catch (final FileAlreadyExistsException ex) {
-                    // We ignore and retry this exception
-                } catch (final Exception ex) {
-                    // Any other exception is bad and we may need to quash.
-                    throw new AssertionError("Unable to create temporary file");
+    /**
+     * Unfortunately, we cannot actually use Files.createTempFile, because that internally depends on
+     * SecureRandom, which results in a circular dependency.
+     */
+    private static synchronized Path createPrivateTmpDir(final String prefix) throws IOException {
+        final Path systemTempDir = Paths.get(System.getProperty("java.io.tmpdir"));
+        if (!Files.isDirectory(systemTempDir)) {
+            throw new AssertionError("java.io.tmpdir is not valid: " + systemTempDir);
+        }
+
+        final byte[] rndBytes = bootstrapRng(Long.BYTES); // Default java tmp files use this much entropy
+        final StringBuilder privateTempDir = new StringBuilder(prefix);
+
+        for (byte b : rndBytes) {
+            // We convert to an unsigned integer first to avoid sign-bit extension when converting to hex.
+            String hexByte = Integer.toHexString(Byte.toUnsignedInt(b));
+            if (hexByte.length() == 1) {
+                privateTempDir.append('0');
+            }
+            privateTempDir.append(hexByte);
+        }
```

Comment:
do we care about uniqueness, unpredictability, or both here?

if all we care about is uniqueness, we could do something stupid like current millis time + JVM pid or something instead of taking the dependency on `/dev/urandom`

Reply:
Unpredictabity is better and we should do as best we can. In the windows world we will probably need to settle on current millis combined with PID and nanotime (though how to combine, I don't know).

Reply:
We should aim for cryptographically strong unpredictability. My understanding is that we should be resilient against an attacker that:
1. Is on the same host, with the privileges of a normal user
2. Can read/write to the same system temporary directory (`/tmp/*`) that ACCP is using
3. Can time their reads/writes with very high precision and throughput

If we only use currentTime + PID we don't have unpredictability, then an attacker could theoretically still mount a DDOS attack on AACP by creating empty directories with the same names that ACCP is attempting to create.

My idea (that I'm not sure would work), would be to have `bootstrapRng()` attempt to use the SunJCE SecureRandom implementation for bootstrapping, but I'm not sure if we'll run into issues where ACCP is being initialized and installed before the SunJCE is ready or if it is not available.

Reply:
For context, when [conscrypt writes out its shared object](https://github.com/google/conscrypt/blob/ca6485b689219323e91d3bbe0410715b5b2d687f/openjdk/src/main/java/org/conscrypt/NativeLibraryLoader.java#L189-L199) its temporary files are [based entirely on the current time](https://github.com/google/conscrypt/blob/ca6485b689219323e91d3bbe0410715b5b2d687f/openjdk/src/main/java/org/conscrypt/Platform.java#L115-L144). While I believe that we can do better, they're solution is likely sufficient.

Reply:
I'll leave the question for what to do when we don't have `/dev/urandom` for a later time. We don't need to answer it until we pick up https://github.com/corretto/amazon-corretto-crypto-provider/issues/48